helix-tui/src/buffer.rs

```diff
@@ -583,6 +583,50 @@ impl Buffer {
         }
         updates
     }
+
+    /// Apply shade to all cells in area.
+    ///
+    /// shade = 0 -> set text modifier DIM
+    /// shade < 0 -> darken rgb color
+    /// shade > 0 -> lighten rgb color
+    pub fn dim(&mut self, area: Rect, shade: i8) {
+        let shade_amount = shade.unsigned_abs().saturating_mul(2);
+        let shade_fn = if shade < 0 {
+            u8::saturating_sub
+        } else {
+            u8::saturating_add
+        };
+        // TODO: replace with saturating_add_signed when mixed_integer_ops (https://github.com/rust-lang/rust/issues/87840) get stabilized.
+        let shaded = |v| shade_fn(v, shade_amount);
+
+        let mut cache: HashMap<Color, Color> = HashMap::new();
+        let mut modify = |col| {
+            let result = cache.entry(col).or_insert_with(|| {
```

Comment:
I suspect that the color transformation is faster than SipHash 2-4 hashing + memory allocation.

Reply:
i just benched it. you are right. way too much overhead at these tiny resolutions.
thanks for catching this. removed it in 8625892af12c3f92e4f6035835169ace0f95e44c